python/cudf/cudf/core/scalar.py

```diff
@@ -1,26 +1,208 @@
+            )
+        return value, dtype
+    elif isinstance(dtype, StructDtype):
+        if value not in {None, NA}:
+            raise ValueError(f"Can not coerce {value} to StructDType")
+        else:
+            return NA, dtype
+
+    if isinstance(dtype, cudf.core.dtypes.DecimalDtype):
+        value = pa.scalar(
+            value, type=pa.decimal128(dtype.precision, dtype.scale)
+        ).as_py()
+    if isinstance(value, decimal.Decimal) and dtype is None:
+        dtype = cudf.Decimal128Dtype._from_decimal(value)
+
+    value = to_cudf_compatible_scalar(value, dtype=dtype)
+
+    if dtype is None:
+        if not valid:
+            if value is NaT:
+                value = value.to_numpy()
+
+            if isinstance(value, (np.datetime64, np.timedelta64)):
+                unit, _ = np.datetime_data(value)
+                if unit == "generic":
+                    raise TypeError("Cant convert generic NaT to null scalar")
+                else:
+                    dtype = value.dtype
+            else:
+                raise TypeError(
+                    "dtype required when constructing a null scalar"
+                )
+        else:
+            dtype = value.dtype
+
+    if not isinstance(dtype, cudf.core.dtypes.DecimalDtype):
+        dtype = cudf.dtype(dtype)
+
+    if not valid:
+        value = NaT if dtype.kind in "mM" else NA
+
+    return value, dtype
+
+
+def _replace_nested(obj, check, replacement):
+    if isinstance(obj, list):
+        for i, item in enumerate(obj):
+            if check(item):
+                obj[i] = replacement
+            elif isinstance(item, (dict, list)):
+                _replace_nested(item, check, replacement)
+    elif isinstance(obj, dict):
+        for k, v in obj.items():
+            if check(v):
+                obj[k] = replacement
+            elif isinstance(v, (dict, list)):
+                _replace_nested(v, check, replacement)
+
+
+def _maybe_nested_pa_scalar_to_py(pa_scalar: pa.Scalar) -> Any:
+    """
+    Convert a "nested" pyarrow scalar to a Python object.
+
+    These scalars come from pylibcudf.Scalar where field names can be
+    duplicate empty strings.
+
+    Parameters
+    ----------
+    pa_scalar: pa.Scalar
+
+    Returns
+    -------
+    Any
+        Python scalar
+    """
+    if not pa_scalar.is_valid:
+        return pa_scalar.as_py()
+    elif pa.types.is_struct(pa_scalar.type):
+        return {
+            str(i): _maybe_nested_pa_scalar_to_py(val)
+            for i, (_, val) in enumerate(pa_scalar.items())
+        }
+    elif pa.types.is_list(pa_scalar.type):
+        return [_maybe_nested_pa_scalar_to_py(val) for val in pa_scalar]
+    else:
+        return pa_scalar.as_py()
+
+
+def _to_plc_scalar(value: ScalarLike, dtype: Dtype) -> plc.Scalar:
+    """
+    Convert a value and dtype to a pylibcudf Scalar for device-side cudf.Scalar
+
+    Parameters
+    ----------
+    value: Scalarlike
+    dtype: dtypelike
```

Comment:
At some point we should really audit functions like this that are mostly internal to see if we can guarantee that we only get actual dtypes rather than something that ducktypes as a dtype. The `pd.api.types` functions are shockingly expensive, especially if they get used in deeper parts of the code like this that get called many times.

Reply:
Sure thing. I think this follows https://github.com/rapidsai/cudf/issues/12494 too